lms/services/h_group.py

```diff
@@ -0,0 +1,79 @@
+from sqlalchemy.dialects.postgresql import insert
+
+from lms.models import HGroup
+from lms.models._hashed_id import hashed_id
+
+MAX_GROUP_NAME_LENGTH = 25
+
+
+class HGroupService:
+    def __init__(self, db):
+        self._db = db
+
+    def upsert(self, name: str, authority_provided_id: str, group_type: str) -> HGroup:
+        upsert_statement = (
+            insert(HGroup)
+            .values(
+                name=name, authority_provided_id=authority_provided_id, type=group_type
+            )
+            .on_conflict_do_nothing(
+                constraint=HGroup.UNIQUE_CONSTRAINT_NAME,
+            )
+        )
+        self._db.execute(upsert_statement)
+        return (
+            self._db.query(HGroup)
+            .filter_by(name=name, authority_provided_id=authority_provided_id)
+            .one()
+        )
```

Comment:
I wonder if we can use the ORM here? It'd be something like:

```python
h_group = self._db.query(HGroup).filter_by(authority_provided_id=...)

if not h_group:
    h_group = HGroup(authority_provided_id=..., type=...)
    self._db.add(h_group)

# Update any fields that might have changed.
h_group.name = name

return h_group
```

Note the distinction between fields that're used to identify the group for looking it up in the DB (`authority_provided_id`) versus fields that we just create/update (`name`)

Reply:
FYI I'm not sure this method implements an upset as written. I think this is a create-if-not-exists. For an upsert you still might need to update certain fields on an pre-existing row in the DB. `name` is the only field that can change I think.

Reply:
Switched to ORM only :+1: